invoice/src/invoice.rs

```diff
@@ -41,13 +44,31 @@ pub enum InvoiceState {
     #[display("")]
     Void,
     #[display("{0}")]
-    Amount(Amount),
-    #[display("...")] // TODO
-    Data(Vec<u8> /* StrictVal */),
+    Amount(CoinAmount),
+    #[display(inner)]
+    Data(NonFungible),
     #[display(inner)]
     Attach(AttachId),
 }
 
+impl FromStr for InvoiceState {
+    type Err = Infallible;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        Ok(if s.is_empty() {
+            InvoiceState::Void
+        } else if let Ok(amount) = CoinAmount::from_str(s) {
+            InvoiceState::Amount(amount)
+        } else if let Ok(data) = NonFungible::from_str(s) {
+            InvoiceState::Data(data)
+        } else if let Ok(attach) = AttachId::from_str(s) {
+            InvoiceState::Attach(attach)
+        } else {
+            unreachable!()
+        })
+    }
+}
```

Comment:
Since this returns an error, I think proper error handling should be used.

```suggestion
#[derive(Clone, PartialEq, Eq, Debug, Display, Error, From)]
#[display(inner)]
pub enum InvoiceStateError {
    #[display(doc_comments)]
    /// could not parse as amount, data, or attach: {0}.
    ParseError(String),
}

impl FromStr for InvoiceState {
    type Err = InvoiceStateError;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        if s.is_empty() {
            Ok(InvoiceState::Void)
        } else if let Ok(amount) = CoinAmount::from_str(s) {
            Ok(InvoiceState::Amount(amount))
        } else if let Ok(data) = NonFungible::from_str(s) {
            Ok(InvoiceState::Data(data))
        } else if let Ok(attach) = AttachId::from_str(s) {
            Ok(InvoiceState::Attach(attach))
        } else {
            Err(InvoiceStateError::ParseError(s.to_owned()))
        }
    }
}
```